File: c/Maximum Number of Words You Can Type.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

int canBeTypedWords(char *text, char *brokenLetters) {
    bool broken[26] = {false};

    // Mark broken letters
    for (int i = 0; brokenLetters[i]; i++) {
        broken[brokenLetters[i] - 'a'] = true;
    }

    int count = 0;
    bool valid = true;

    for (int i = 0; ; i++) {
        char c = text[i];

        if (c == ' ' || c == '\0') {
            if (valid) count++;   // Word was valid
            valid = true;         // Reset for next word
            if (c == '\0') break; // End of string
        } else {
            if (broken[c - 'a']) valid = false; // Word is invalid
        }
    }

    return count;
}
// Test cases
int main() {
    // case 1
    char text1[] = "hello world";
    char broken1[] = "ad";
    printf("Result of test case 1 : %d\n", canBeTypedWords(text1, broken1)); // Expected 1

    // Example 2
    char text2[] = "leet code";
    char broken2[] = "lt";
    printf("Result of test case %d\n", canBeTypedWords(text2, broken2)); // Expected 1

    // Example 3
    char text3[] = "leet code";
    char broken3[] = "e";
    printf("Result of test case 3 : %d\n", canBeTypedWords(text3, broken3)); // Expected 0

    return 0;
}

File: JavaCode/MaximumNumberofWordsYouCanType.java
public class MaximumNumberofWordsYouCanType {
    public static void main(String[] args) {
        Solution sol = new Solution();

        // case 1
        String text1 = "hello world";
        String brokenLetters1 = "ad";
        System.out.println("Result of test case 1 : " + sol.canBeTypedWords(text1, brokenLetters1)); // Expected 1

        // case 2
        String text2 = "leet code";
        String brokenLetters2 = "lt";
        System.out.println("Result of test case 2 : " + sol.canBeTypedWords(text2, brokenLetters2)); // Expected 1

        // case 3
        String text3 = "leet code";
        String brokenLetters3 = "e";
        System.out.println("Result of test case 3 : " + sol.canBeTypedWords(text3, brokenLetters3)); // Expected 0
    }
}

class Solution {
    public int canBeTypedWords(String text, String brokenLetters) {
        boolean[] broken = new boolean[26];
        for (char c : brokenLetters.toCharArray()) {
            broken[c - 'a'] = true;
        }

        int count = 0;
        boolean valid = true;

        for (int i = 0; i <= text.length(); i++) {
            char c = (i == text.length()) ? '\0' : text.charAt(i);

            if (c == ' ' || c == '\0') {
                if (valid) count++;
                valid = true; 
                if (c == '\0') break;
            } else {
                if (broken[c - 'a']) valid = false;
            }
        }

        return count;
    }
}

